### PR TITLE
RED-24: `TransitionMatrix` and `DefineStationaryDist` animations

### PR DESCRIPTION
**Pull request by jesi-rgb**
## Problem
`TransitionMatrix` and `DefineStationaryDist` needed to be animated

## Solution
Implementation of the animation for both scenes. 

- `TransitionMatrix` showcases how the different edges of the chain have different probabilities and later, a text mob shows these probabilities being calculated for an arbitrary number (5) of steps.
- `DefineStationaryDist` then presents the concept of a stationary distribution with mathematical notation as well as a small simulation of the same markov chain where the labels present the corresponding distribution. One can appreciate how these labels converge fairly quickly. Then, two questions are presented enhancing the script.

## Test
Animations run as expected. Here's a quick, shortened preview:

https://user-images.githubusercontent.com/50735312/166140356-0b63eea1-9c04-480f-babb-bcd5aaba0359.mp4

### CODE REVIEW

**Comment by jesi-rgb** [comment]
drafting this while waiting for feedback! it'll probably need more revisions @nipunramk

**Comment by nipunramk** [comment]
@jesi-rgb here are my suggestions for this animation:
1. I added more explanation in the script based on your animation so be sure to reference that
2.  At 0:05, the arrow from (1, 2) should have opacity reduced
3. Let's maybe build up to the equations in 0:07 (see script for how I explain it)
- We can start by saying we want to calculate pi_{n + 1}(0) which requires knowing pi_{n}(0),  pi_{n}(1),  pi_{n}(2),  pi_{n}(3), and  pi_{n}(4) (the probabilities of being in each state at the previous step.
- Then introduce the idea of a weighted sum across all previous states multiplied by P(i, 0) where is all possible previous states (so i = {0, 1, 2, 3, 4})
- Now you have a weighted sum of 5 elements, and now you can directly show this as a dot product between row vector pi_n and a column vector with elements [P(0, 0), P(1, 0), P(2, 0), P(3, 0), P(4, 0)]
- And after this you can introduce the concept of the transition matrix
- Also as a small delta: let's use \cdot to show multiplication rather than *
4. At 0:14, you show the transition matrix -- could you also put the nodes at the top of the transition matrix to label the columns as you do with the rows. The row labels look great. Also, could we make the matrix with brackets?
5. From 0:14 - 0:24, we do not need to show this so remove this section, only show one transition matrix (labeled as P). It is not obvious what this means and I know what you mean, but it gets too confusing. In practice, Markov chains do not even calculate these intermediate matrices, since matrix exponentiation can be expensive. As we discussed, the main way to show transitions is by multiplying pi with P and updating the distributions step by step. We never actually calculate P^2, P^3, etc. 
6. At 0:28, define a stationary distribution pi (no steps associated with it), and then the actual mathematical definition is that pi = pi * P (equation 1.1 in Probability in Electrical Engineering book).

**Comment by jesi-rgb** [comment]
@nipunramk tried my best at adding everything you told me to the scene! hopefully i didn't miss anything. unless i missed it, it should be ready to be merged!

here's a preview! the labels are not positioned correctly because i've not been able to merge those changes with this branch, but wil surely look good on your end!


https://user-images.githubusercontent.com/50735312/166744204-4430dbab-84cf-483d-905c-e3d6f61e22dc.mp4

**Comment by nipunramk** [comment]
@jesi-rgb 
Looking good!
Couple more suggestions:

0:21, this may not be strictly necessary but I think it might be nice to have one more mathematical expression showing this particular line in the script: " If you think about this calculation, it’s essentially a dot product!"

After 0:25, I think we are missing some animations for this part of the script:
"Now, what’s really nice about Markov chains is we can find the probabilities of landing in each state at the next step really easily: it’s just a matrix vector product of the probabilities at the previous state times the transition matrix. The convention used in Markov chains is the state probability vectors are represented as row vectors."

Showing pi_n+1 = pi_n * P might be a good idea and defining pi_n in general as a row vector is something we're going to use the rest of the video, so good to introduce here. 

And lastly, maybe a couple more steps of the user simulation may be necessary to fill the space in the script. (maybe n = 15 where the last 7 or so have the distributions frozen). Again we can always speed up in post-processing. 

In general, my feedback on your animations is to be very meticulous about following the script. Try to make sure every sentence is accounted for in context of a larger animation or in its own specific animation. I am being this meticulous with it because the more we take the time to do this part right, post-processing/editing will be significantly faster.

**Comment by jesi-rgb** [comment]
@nipunramk This is a first version with lots of math notation added. Feel free to absolutely roast anything at all. I tried to follow the script as best as I could, but let me know if some parts could go better in other places

https://user-images.githubusercontent.com/50735312/166909967-4c358c3e-83ac-4e6f-b13d-7b4c42522adb.mp4

**Comment by jesi-rgb** [comment]
Merged all your changes to this branch so there should not be any problems when bringing it back to your computer!

**Comment by jesi-rgb** [comment]
Addressed all the feedback and added the animation I told you. I think it works great and makes the concept click a bit more. Also, it's just beautiful!

Feel free to absolutely roast and obliterate the math notation part. We want to make sure that is not only correct, but well laid out, @nipunramk 

Lemme know your thoughts!

https://user-images.githubusercontent.com/50735312/166921325-d7975cfe-ec32-4469-8b1d-459f0bd84946.mp4

**Comment by nipunramk** [comment]
@jesi-rgb  looks incredible. Only one minor change: at 0:24, the vector hat symbol is not aligned to pi, just change that and we should be good to merge. Let me know proper ordering of the merge. Same issue btw at 0:30 with the location of the vector symbol above P.

File: PageRank/jesus_animations.py [hunk 1]
```diff
@@ -34,31 +35,449 @@ def construct(self):
...
+            .scale(0.5)
+            .move_to(frame.get_corner(UL), aligned_edge=UL)
+            .shift(DR * 0.5)
+        )
+        self.play(FadeIn(math_notation_title, shift=UP * 0.3), FadeOut(full_equation))
+
+        dist_definition = (
+            MathTex(
+                # r"\vec{\pi_n} = [\pi_n(0), \pi_n(1), \pi_n(2), \pi_n(3), \pi_n(4) ]",
+                r"\vec{\pi_n} = \begin{bmatrix} \pi_n(0) & \pi_n(1) & \pi_n(2) & \pi_n(3) & \pi_n(4) \end{bmatrix}",
```
Comment (nipunramk) [inline PageRank/jesus_animations.py]: \vec{\pi}_n should solve the problem

File: PageRank/jesus_animations.py [hunk 2]
```diff
@@ -34,31 +35,449 @@ def construct(self):
...
+
+        self.play(FadeIn(dist_definition, shift=UP * 0.3))
+        self.wait()
+
+        self.play(dist_definition.animate.shift(UP * 2))
+        self.wait()
+
+        trans_column_def = (
+            MathTex(
+                r"\vec{P_{i,0}} = \begin{bmatrix} P(0,0) \\ P(1,0) \\ P(2,0) \\ P(3,0) \\ P(4,0) \end{bmatrix}"
```
Comment (nipunramk) [inline PageRank/jesus_animations.py]: \vec{P}_{i, 0} should solve the problem

**Pull request merged**